### ISSUE TITLE
[Bug Report] Matmul 1d/2d non-deterministic output and hangs when run in a loop

### ISSUE THREAD

**Issue by s-jovic** [tenstorrent/tt-metal#7066]
**Describe the bug**
We observe non-deterministic output of falcon_lm_head_matmul for 7b when we run it in a loop with random inputs.

**To Reproduce**
```
from loguru import logger

import tt_lib as ttl
from models.utility_functions import comp_pcc, tt2torch_tensor
import torch

def test_reproduce_lm_head_nd_32(
    device,
):
    in0_mem_config = ttl.tensor.MemoryConfig(ttl.tensor.TensorMemoryLayout.INTERLEAVED, ttl.tensor.BufferType.L1)
    in1_mem_config = ttl.tensor.MemoryConfig(ttl.tensor.TensorMemoryLayout.INTERLEAVED, ttl.tensor.BufferType.DRAM)
    out_mem_config = ttl.tensor.MemoryConfig(ttl.tensor.TensorMemoryLayout.INTERLEAVED, ttl.tensor.BufferType.L1)

    in0_dtype = ttl.tensor.DataType.BFLOAT16
    in1_dtype = ttl.tensor.DataType.BFLOAT8_B
    out_dtype = ttl.tensor.DataType.BFLOAT16

    torch.manual_seed(1234)

    seq_len = 32
    a_shape = [1, 1, seq_len, 4544]
    b_shape = [1, 1, 4544, 65024]

    A = torch.randn(a_shape)
    B = torch.randn(b_shape) - 0.95

    a_t = ttl.tensor.Tensor(A, in0_dtype).to(ttl.tensor.Layout.TILE).to(device, in0_mem_config)
    b_t = ttl.tensor.Tensor(B, in1_dtype).to(ttl.tensor.Layout.TILE).to(device, in1_mem_config)
    bias_t = None

    out = ttl.tensor.falcon_lm_head_matmul(a_t, b_t, bias_t, output_mem_config=out_mem_config, output_dtype=out_dtype)

    ref_out = tt2torch_tensor(out)

    nd_output_count = 0

    for _ in range(100):
        out.deallocate(True)
        out = ttl.tensor.falcon_lm_head_matmul(a_t, b_t, bias_t, output_mem_config=out_mem_config, output_dtype=out_dtype)

        pt_out = tt2torch_tensor(out)

        _, output_pcc = comp_pcc(ref_out, pt_out, 1)

        if output_pcc != 1:
            nd_output_count += 1

        logger.debug(f"Output pcc={output_pcc}")

    print(f"Iterations with nd output: {nd_output_count}")
```

**Additional context**
- We get nd output ~13 times when running a loop of 100 runs
- If we disable l1 accumulation, it drops to a couple of times in 1000 runs
- slow dispatch doesn't help
- running with 500Hz helps ([idea from this issue](https://github.com/tenstorrent-metal/tt-metal/issues/4968#issuecomment-1930663733), which seems to be related)
- other seq lens have deterministic output
- @skhorasganiTT noticed same behavior with mlp h_t_4h matmul which is also matmul 1D ([issue](https://github.com/tenstorrent-metal/tt-metal/issues/6795), this is probably affecting nd 7b demo hang, so putting P1 critical)

**Comment by s-jovic** [comment]
Additional info, I tried changing L1 interleaved for activations and output to DRAM interleaved, and I see a hang after some iterations when L1 accumulation is also enabled. To sum up:
- activations & output in L1 interleaved + L1 acc enabled -> frequent nd output
- activations & output in L1 interleaved + L1 acc disabled -> less frequent nd output but still there
- activations & output in DRAM + L1 acc enabled -> nd output + **hang at some point**
- activations & output in DRAM + L1 acc disabled -> nd output, but no hang in 10k iterations

**Comment by s-jovic** [comment]
Falcon40b is affected by this bug as well - lm head and first MLP matmul with 32/128 sequence lengths. Again, fixed if subblock h/w are set to 1, or grid size changed. This is not blocking for 40b, when we set subblock h/w to 1, we don't see meaningful perf regression. 7b lm_head originally posted in the issue can also be fixed to use subblock h/w = 1 without regressing perf meaningfully.

Sending the script to reproduce issues with 40b matmuls in case it is useful.

```
from loguru import logger
import pytest

import tt_lib as ttl
from models.utility_functions import comp_pcc, tt2torch_tensor
import torch

@pytest.mark.parametrize(
    "seq_len, weights_n, per_core_M, per_core_N, in_block_w, out_subblock_h, out_subblock_w, loop_count",
    (
        (32, 8128, 1, 4, 4, 1, 4, 1000), # lm head
        (128, 8128, 4, 4, 4, 1, 4, 2000), # lm head
        (32, 4096, 1, 2, 4, 1, 2, 2000), # mlp 4h
        (128, 4096, 4, 2, 4, 2, 2, 20000), # mlp 4h
    ),
    ids=["lm_seq_len32", "lm_seq_len128", "mlp_4h_seq_len32", "mlp_4h_seq_len128"]
)
def test_reproduce_matmul_1d(
    device,
    seq_len,
    weights_n,
    per_core_M,
    per_core_N,
    in_block_w,
    out_subblock_h,
    out_subblock_w,
    loop_count
):
    in0_mem_config = ttl.tensor.MemoryConfig(ttl.tensor.TensorMemoryLayout.INTERLEAVED, ttl.tensor.BufferType.DRAM)
    in1_mem_config = ttl.tensor.MemoryConfig(ttl.tensor.TensorMemoryLayout.INTERLEAVED, ttl.tensor.BufferType.DRAM)
    out_mem_config = ttl.tensor.MemoryConfig(ttl.tensor.TensorMemoryLayout.INTERLEAVED, ttl.tensor.BufferType.DRAM)

    in0_dtype = ttl.tensor.DataType.BFLOAT8_B
    in1_dtype = ttl.tensor.DataType.BFLOAT8_B
    out_dtype = ttl.tensor.DataType.BFLOAT8_B

    torch.manual_seed(1234)

    a_shape = [1, 1, seq_len, 8192]
    b_shape = [1, 1, 8192, weights_n]

    A = torch.randn(a_shape)
    B = torch.randn(b_shape) - 0.95

    a_t = ttl.tensor.Tensor(A, in0_dtype).to(ttl.tensor.Layout.TILE).to(device, in0_mem_config)
    b_t = ttl.tensor.Tensor(B, in1_dtype).to(ttl.tensor.Layout.TILE).to(device, in1_mem_config)
    bias_t = None

    program_config = ttl.operations.primary.MatmulMultiCoreReuseMultiCast1DProgramConfig(
        compute_with_storage_grid_size=(8, 8),
        in0_block_w=in_block_w,
        out_subblock_h=out_subblock_h,
        out_subblock_w=out_subblock_w,
        per_core_M=per_core_M,
        per_core_N=per_core_N,
        fuse_batch=True,
        fused_activation=None,
        mcast_in0=True,
    )

    compute_config = ttl.tensor.WormholeComputeKernelConfig(
        math_fidelity=ttl.tensor.MathFidelity.LoFi,
        math_approx_mode=True,
        fp32_dest_acc_en=True,
        packer_l1_acc=False,
    )

    out = ttl.operations.primary.matmul_1d(
        a_t,
        b_t,
        program_config=program_config,
        output_mem_config=out_mem_config,
        output_dtype=out_dtype,
        compute_kernel_config=compute_config
    )

    ref_out = tt2torch_tensor(out)

    nd_output_count = 0

    for _ in range(loop_count):
        out.deallocate(True)

        out = ttl.operations.primary.matmul_1d(
            a_t,
            b_t,
            program_config=program_config,
            output_mem_config=out_mem_config,
            output_dtype=out_dtype,
            compute_kernel_config=compute_config
        )

        pt_out = tt2torch_tensor(out)

        _, output_pcc = comp_pcc(ref_out, pt_out, 1)

        if output_pcc != 1:
            nd_output_count += 1

        logger.debug(f"Output pcc={output_pcc}")

    print(f"Iterations with nd output: {nd_output_count}")

    assert nd_output_count != 0
```

**Comment by s-jovic** [comment]
I encountered similar behavior with matmul 2d (nd output + hang):

```
    a_shape = [1, 1, 2048, 8192]
    b_shape = [1, 1, 8192, 1024]

    program_config = ttl.operations.primary.MatmulMultiCoreReuseMultiCastProgramConfig(
        compute_with_storage_grid_size=(8, 8),
        in0_block_w=16,
        out_subblock_h=1,
        out_subblock_w=4,
        per_core_M=8,
        per_core_N=4,
        fused_activation=None,
        transpose_mcast=False,
    )
```

Setting subblock h/w to 1 fixes the issue, however using watcher doesn't help this time.

**Comment by jliangTT** [comment]
Spoke to @uaydonat and he mentioned that this issue is non-blocking for this falcon7B demo at the moment. We will downgrade this to p1 for now as we continue to investigate.

**Comment by smehtaTT** [comment]
FYI from 4/10 Javelin SW bug review - The workaround was to use 8x7 instead of 8x7, so we only use 56 cores

**Comment by uaydonat** [comment]
@s-jovic we have been asked to try hanging tests on the fast dispatch branch to see if FD fixes it. Currently only single-chip support is available, I think the matmul at the top would be a good candidate to try.

**Comment by uaydonat** [comment]
@aliuTT is this the right branch?

**Comment by rtawfik01** [comment]
Hey guys, here are my findings so far, for both the falcon lm head and the matmul 1d non-determinism.

1. Non-deterministic PCCs only stop occurring when AICLK is in the 500-620Mhz range/0.95V.  Once >620Mhz, non-deterministic PCCs start occurring, and around 900-1000Mhz, hangs start occurring.
2. Once stalls/cycles are  added on the unpacker thread in the matmul loop:
```
                    for (uint32_t inner_dim_idx = 0; inner_dim_idx < in0_block_w; ++inner_dim_idx) {
                        // matmul outer product of (out_subblock_h x out_subblock_w) tiles that fill dst
                        // accumulation is done by iterating matmul_block across inner dim
                        // in0_block_w is passed as innder dim (kt) to matmul_block, interally used to stride in0
                        matmul_block(in0_cb_id, in1_cb_id, in0_index, in1_index, dst_index, false, out_subblock_w, out_subblock_h, in0_block_w);
                        in0_index ++;  // stride right by 1
                        in1_index += in1_per_core_w; // to stride down by 1 need to stride by in_per_core_w (should be called in1_block_w)
                    }
```
The hang/non-determinism stops occurring, If the stalls are added elsewhere in the kernel, the issue still persists.

3. For the matmul 1d hang, the issue stops occurring specifically when stalling `TTI_STALLWAIT(p_stall::STALL_UNPACK, p_stall::UNPACK0);` in the matmul block loop, adding a stallwait on the other unpack instance UNPACK1, still causes non-deterministic issues.

In summary, since the issue occurs for an AICLK that is lower than what was required to fix the di/dt, this might be a real issue occurring from a thread not completing while executing new configs. I'll keep investigating further and update if there is a fix.

**Comment by aliuTT** [comment]
Yes, `fd-2/main` is the most up to date FD2 branch.

**Comment by davorchap** [comment]
related case #7168 that @TT-billteng and @yugaoTT found on CI, but no repro locally so far

**Comment by s-jovic** [comment]
@ttmtrajkovic @rtawfik01 

Hey guys, you can now reproduce a hang on main with this command:
`pytest models/experimental/falcon_40b/tests/test_reproduce_nd_matmul.py::test_reproduce_matmul_2d -k "mlp_4h_seq_len2048"`

It hangs in less than 100 iterations, and exposes non-deterministic output as well.

**Comment by rtawfik01** [comment]
> pytest models/experimental/falcon_40b/tests/test_reproduce_nd_matmul.py::test_reproduce_matmul_2d -k "mlp_4h_seq_len2048"

Hey @s-jovic I am unable to reproduce this hang or non-determinism, can you tell me which commit on main you were using? I am on this commit: f0700725d

**Comment by pavlejosipovic** [comment]
@rtawfik01 I've tried on latest main and the test case above was hard to repro (I was lucky to repro ND just once out of many times).
On the other hand second and third variants of the test consistently had ND behaviour.
`pytest models/experimental/falcon_40b/tests/test_reproduce_nd_matmul.py::test_reproduce_matmul_2d -k "mlp_h_seq_len2048"` 
`pytest models/experimental/falcon_40b/tests/test_reproduce_nd_matmul.py::test_reproduce_matmul_2d -k "attn_seq_len2048"`
I wasn't able to repro any hangs.
(I was testing on single card N150 at 800Mhz)

**Comment by pavlejosipovic** [comment]
It turns out that running same test on a single card but on t3k system will trigger hang easily. 
Also, when run on N150 with new yet unreleased firmware which bumps the voltage issues goes away.

**Comment by pavlejosipovic** [comment]
On N150 if `use_program_cache` is added to the test `mlp_h_seq_len2048` and `attn_seq_len2048` variants consistently produce a hang.
If voltage is bumped with version with `use_program_cache` works as expected.

**Comment by jliangTT** [comment]
Update: 
- attempted repro two [weeks](https://github.com/tenstorrent/tt-metal/issues/7066#issuecomment-2064222268) ago, should attempt repro again.

**Comment by rtawfik01** [comment]
Here is the summary of the tests run for this issue, for the initial falcon 7b and matmul_1d non deterministic pcc:

1. If the AICLK is lowered to 500-620Mhz range, from the initial 800Mhz, the PCC becomes deterministic ==> This indicates a timing issue
2. If STALLWAITS are added on UNPACK0, which slows down the unpack operation, the PCC also becomes deterministic  ==> This indicates a timing issue
3. When the voltage margin was bumped up, the PCC also became deterministic ==> This indicates a timing issue due to a potential di/dt event.

Syseng team is going to try this test and update if it is a di/dt event or another potential issue. In the meantime @pavlejosipovic @s-jovic it will be good to try to get a test that reproduces consistent hangs on a single chip, I was unable to get a lot of ND hangs.

**Comment by uaydonat** [comment]
@rtawfik01 when you say "When the voltage margin was bumped up," is this just running it with the new fw that released that bumped up the clock to 1GHz as well? Or did you try a different fw that just increases the voltage but not the clock?

**Comment by ttmtrajkovic** [comment]
@uaydonat,
Both have been tried. Controlled experiment on Reem's machine was done just to increase the margin. 

@pavlejosipovic tested with both margin and 1GHz clock (basically, new FW) and got the same results, so seems like new FW is fixing these problems as well, although its not super intuitive why as pcc randomness and very low passing point (620MHz) are unusual symptoms for a di/dt event (at least to me).
I am taking this conversation further with syseng team that's more knowledgable of this so I will update when more info is there.

Milos

**Comment by uaydonat** [comment]
@pavlejosipovic we should try this with FD2 on latest main to see if the hang is gone.
